File: packages/js/lib/coerce.js
// Vector params (interpolated plain JSON, AWS API-model member names) →
// aws-sdk-js-v3 command inputs. v3 has no runtime shape metadata, so the
// conversions the SDK's serializers require are driven by explicit,
// corpus-derived key sets; the offline corpus smoke test is the drift alarm
// when a corpus bump introduces new shapes.

import { contentValue } from './match.js'

// Fields (at any depth) whose values are timestamps the serializers need as
// Date instances.
const TIMESTAMP_KEYS = new Set([
  'CopySourceIfModifiedSince', 'CopySourceIfUnmodifiedSince',
  'IfModifiedSince', 'IfUnmodifiedSince',
  'IfMatchLastModifiedTime', 'Expires',
  'ObjectLockRetainUntilDate', 'RetainUntilDate',
  'LastModifiedTime',
  'Date'
])

// Fields modeled as strings that the corpus writes as bare numbers.
const STRINGIFY_KEYS = new Set(['PartNumberMarker'])

/**
 * Build a v3 command input from interpolated vector params.
 * @param {object} params
 * @param {(name: string) => Uint8Array} resolveData $data resolver
 * @param {(name: string) => ({stream: import('node:stream').Readable, length: number} | null)} [resolveStream]
 *   optional: when it returns a stream for a bare {"$data": name} Body, the
 *   body is handed to the SDK as a stream instead of being materialized, so a
 *   multi-gigabyte body costs one chunk rather than its own size. Returning
 *   null means "materialize this one", which is how small datasets stay on the
 *   cached-bytes path. Everything else always resolves through resolveData.
 * @returns {{input: object, body: Uint8Array | null, contentLength: number | null}}
 *   body = held-aside Body bytes (the presign path sends them itself), null
 *   when the body was streamed
 */
/** The dataset name of a bare {"$data": name} descriptor, else null. */
function dataRef (value) {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return null
  if (Object.keys(value).length !== 1 || typeof value.$data !== 'string') return null
  return value.$data
}

export function buildInput (params, resolveData, resolveStream) {
  let body = null
  let contentLength = null
  const walk = (value, key) => {
    if (key === 'Body') {
      const name = dataRef(value)
      if (resolveStream && name !== null) {
        const s = resolveStream(name)
        if (s) {
          contentLength = s.length
          return s.stream
        }
      }
      body = contentValue(value, resolveData)
      return body
    }
    if (typeof value === 'string') {
      if (TIMESTAMP_KEYS.has(key)) return parseTime(value)
      return value
    }
    if (typeof value === 'number' && STRINGIFY_KEYS.has(key)) return String(value)
    if (Array.isArray(value)) return value.map((e) => walk(e, key))
    if (value !== null && typeof value === 'object') {
      // Object-valued string params: boto3-style CopySource and binary
      // content descriptors (SSE-C keys), whose wire form is base64.
      if (key === 'CopySource' && typeof value.Bucket === 'string' && typeof value.Key === 'string') {
        // The SDK does not encode CopySource, so percent-encode the key here
        // (each segment; '/' separators are preserved) to keep special
        // characters (spaces, '?', ...) out of the wire header.
        const encKey = value.Key.split('/').map(encodeURIComponent).join('/')
        let src = value.Bucket + '/' + encKey
        if (typeof value.VersionId === 'string' && value.VersionId !== '') src += '?versionId=' + value.VersionId
        return src
      }
      if (('$base64' in value || '$data' in value) && Object.keys(value).length === 1) {
        return Buffer.from(contentValue(value, resolveData)).toString('base64')
      }
      const out = {}
      for (const [k, v] of Object.entries(value)) out[k] = walk(v, k)
      return out
    }
    return value
  }
  const input = {}
  for (const [k, v] of Object.entries(params)) input[k] = walk(v, k)
  // A streamed body has no length the SDK can discover, so supply it — unless
  // the vector set ContentLength itself.
  if (contentLength !== null && input.ContentLength === undefined) input.ContentLength = contentLength
  return { input, body, contentLength }
}

const RFC3339 = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(\.\d+)?(Z|[+-]\d{2}:\d{2})$/
const HTTP_DATE = /^[A-Z][a-z]{2}, \d{2} [A-Z][a-z]{2} \d{4} \d{2}:\d{2}:\d{2} GMT$/
const BARE_DATE = /^(\d{4})-(\d{2})-(\d{2})$/
const COMPACT_DATE = /^(\d{4})(\d{2})(\d{2})$/

/**
 * Parse the corpus's timestamp shapes: RFC3339 (the norm), HTTP-date
 * ("Expires" params) and bare dates (lifecycle rules). Throws on anything
 * else (a vector-definition error → RunnerError).
 * @param {string} s
 * @returns {Date}
 */
export function parseTime (s) {
  let m
  if (RFC3339.test(s) || HTTP_DATE.test(s)) {
    const d = new Date(s)
    if (!Number.isNaN(d.getTime())) return d
  } else if ((m = s.match(BARE_DATE)) || (m = s.match(COMPACT_DATE))) {
    return new Date(Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3])))
  }
  throw new Error(`unparseable timestamp ${JSON.stringify(s)}`)
}

File: packages/python/src/cloud_portable_s3tests/_coerce.py
"""Vector params (interpolated plain JSON, AWS API-model member names) →
boto3 call arguments. botocore carries the service model at runtime, so
coercion is driven by each member's declared shape: timestamps become
datetimes, the streaming ``Body`` blob becomes bytes (held aside for the
presign path), binary content descriptors for string members become base64,
numbers written for string members become strings."""

from __future__ import annotations

import base64
import json
import re
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Any, Optional

from ._match import Resolver, content_value


class CoerceError(ValueError):
    """A vector-definition problem (unparseable timestamp, bad content)."""


def build_input(operation_model, params: dict[str, Any], resolve: Optional[Resolver]) -> tuple[dict[str, Any], Optional[bytes]]:
    """Coerce interpolated vector params for ``operation_model``. Returns the
    call kwargs and the held-aside Body bytes (None when the operation sends
    no body)."""
    body: Optional[bytes] = None
    shape = operation_model.input_shape if operation_model is not None else None

    def walk(value: Any, member_shape, key: str) -> Any:
        nonlocal body
        if key == "Body" and member_shape is not None and member_shape.type_name == "blob":
            body = content_value(value, resolve)
            return body
        if key == "CopySource" and isinstance(value, dict) and isinstance(value.get("Bucket"), str) and isinstance(value.get("Key"), str):
            # boto3-style object form; composed like the Go and JS runners
            # (verbatim, no URL-escaping — the corpus pre-encodes).
            src = value["Bucket"] + "/" + value["Key"]
            if isinstance(value.get("VersionId"), str) and value["VersionId"] != "":
                src += "?versionId=" + value["VersionId"]
            return src
        t = member_shape.type_name if member_shape is not None else None
        if t == "timestamp":
            return parse_time(value) if isinstance(value, str) else value
        if t == "string":
            if isinstance(value, dict) and len(value) == 1 and ("$base64" in value or "$data" in value):
                # Binary content descriptors for string members (SSE-C keys)
                # travel base64-encoded.
                return base64.b64encode(content_value(value, resolve)).decode("ascii")
            if isinstance(value, (int, float)) and not isinstance(value, bool):
                return json.dumps(value)  # e.g. PartNumberMarker written as a bare number
            return value
        if t == "blob":
            return content_value(value, resolve) if isinstance(value, (str, dict)) else value
        if t == "structure" and isinstance(value, dict):
            return {k: walk(v, member_shape.members.get(k), k) for k, v in value.items()}
        if t == "list" and isinstance(value, list):
            return [walk(e, member_shape.member, key) for e in value]
        if t == "map" and isinstance(value, dict):
            return {k: walk(v, member_shape.value, k) for k, v in value.items()}
        return value

    members = shape.members if shape is not None else {}
    out = {k: walk(v, members.get(k), k) for k, v in params.items()}
    return out, body


_RFC3339 = re.compile(r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(\.\d+)?(Z|[+-]\d{2}:\d{2})$")
_HTTP_DATE = re.compile(r"^[A-Z][a-z]{2}, \d{2} [A-Z][a-z]{2} \d{4} \d{2}:\d{2}:\d{2} GMT$")
_BARE_DATE = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")
_COMPACT_DATE = re.compile(r"^(\d{4})(\d{2})(\d{2})$")


def parse_time(s: str) -> datetime:
    """Parse the corpus's timestamp shapes: RFC3339 (the norm), HTTP-date
    ("Expires" params) and bare dates (lifecycle rules). Raises on anything
    else (a vector-definition error → RunnerError)."""
    try:
        if _RFC3339.match(s):
            return datetime.fromisoformat(s.replace("Z", "+00:00")).astimezone(timezone.utc)
        if _HTTP_DATE.match(s):
            return parsedate_to_datetime(s).astimezone(timezone.utc)
        m = _BARE_DATE.match(s) or _COMPACT_DATE.match(s)
        if m:
            return datetime(int(m.group(1)), int(m.group(2)), int(m.group(3)), tzinfo=timezone.utc)
    except ValueError:
        pass
    raise CoerceError(f"unparseable timestamp {json.dumps(s)}")
